contrib/hashicorp/vault/vault_test.go

```diff
@@ -0,0 +1,440 @@
+		if err != nil {
+			t.Fatal(err)
+		}
+		// Read key with namespace
+		_, err = client.Logical().Read(key)
+		if err != nil {
+			t.Fatal(err)
+		}
+		spans := mt.FinishedSpans()
+		assert.Len(spans, 2)
+		span := spans[1]
+
+		assert.Equal("vault", span.Tag(ext.ServiceName))
+		assert.Equal(fullPath, span.Tag(ext.HTTPURL))
+		assert.Equal(http.MethodGet, span.Tag(ext.HTTPMethod))
+		assert.Equal(http.MethodGet+" "+fullPath, span.Tag(ext.ResourceName))
+		assert.Equal(ext.SpanTypeHTTP, span.Tag(ext.SpanType))
+		assert.Equal(200, span.Tag(ext.HTTPCode))
+		assert.Zero(span.Tag(ext.Error))
+		assert.Zero(span.Tag(ext.ErrorMsg))
+		assert.Equal(namespace, span.Tag("vault.namespace"))
+	})
+}
+
+func getWriteSpan(t *testing.T, conf *api.Config) mocktracer.Span {
+	assert := assert.New(t)
+	client, err := api.NewClient(conf)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer mountKV(client, t)()
+
+	mt := mocktracer.Start()
+	defer mt.Stop()
+
+	_, err = client.Logical().Write(
+		secretMountPath+"/key",
+		map[string]interface{}{"Key1": "Val1", "Key2": "Val2"},
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	spans := mt.FinishedSpans()
+	assert.Len(spans, 1)
+	return spans[0]
+}
+
+func TestOptionServiceName(t *testing.T) {
+	assert := assert.New(t)
+
+	ts, cleanupTs := setupServer(t)
+	defer cleanupTs()
+
+	t.Run("WithServiceName", func(t *testing.T) {
+		// Check default
+		config := &api.Config{
+			HttpClient: NewHTTPClient(),
+			Address:    ts.URL,
+		}
+		span := getWriteSpan(t, config)
+		assert.Equal(serviceName, span.Tag(ext.ServiceName))
+
+		customServiceName := "someServiceName"
+		config = &api.Config{
+			HttpClient: NewHTTPClient(WithServiceName(customServiceName)),
+			Address:    ts.URL,
+		}
+		span = getWriteSpan(t, config)
+		assert.Equal(customServiceName, span.Tag(ext.ServiceName))
+	})
+
+	t.Run("WithAnalytics", func(t *testing.T) {
+		// With no option, default should be no tag.
+		config := &api.Config{
+			HttpClient: NewHTTPClient(),
+			Address:    ts.URL,
+		}
+		span := getWriteSpan(t, config)
+		// Zero value for type (nil), not 0.0 float value
+		assert.Zero(span.Tag(ext.EventSampleRate))
+
+		// True should set the tag to 1.0
+		config = &api.Config{
+			HttpClient: NewHTTPClient(WithAnalytics(true)),
+			Address:    ts.URL,
+		}
+		span = getWriteSpan(t, config)
+		assert.Equal(1.0, span.Tag(ext.EventSampleRate))
+
+		// false should remove the tag.
+		config = &api.Config{
+			HttpClient: NewHTTPClient(WithAnalytics(false)),
+			Address:    ts.URL,
+		}
+		span = getWriteSpan(t, config)
+		// Zero value for type (nil), not 0.0 float value
+		assert.Zero(span.Tag(ext.EventSampleRate))
+
+		//Last option should win
+		config = &api.Config{
+			HttpClient: NewHTTPClient(WithAnalyticsRate(0.7), WithAnalytics(true)),
+			Address:    ts.URL,
+		}
+		span = getWriteSpan(t, config)
+		assert.Equal(1.0, span.Tag(ext.EventSampleRate))
+	})
+
+	t.Run("WithAnalyticsRate", func(t *testing.T) {
+		//Negative Should be removed.
```

Comment:
There's a few concerns here. We should aim to make unit tests 10x more readable than the actual code. When the code is hard to figure out, unit tests should be a good explanation. On that note, let's make sure a unit test has only one setup section, and one assertion section, and not: setup, assert, setup, assert, etc. That makes it seem more like an integration test.

There is a lot of repetition too, which should pose questions. I would suggest something like this, completely removing repetition:

```go
func TestOption(t *testing.T) {
	ts, cleanup := setupServer(t)
	defer cleanup()

	for _, tt := range []struct {
		opts []Option
		test func(assert *assert.Assertions, span mocktracer.Span)
	}{
		{
			test: func(assert *assert.Assertions, span mocktracer.Span) {
				assert.Equal(serviceName, span.Tag(ext.ServiceName))
				assert.Nil(span.Tag(ext.EventSampleRate))
			},
		},
		{
			opts: []Option{WithAnalytics(true)},
			test: func(assert *assert.Assertions, span mocktracer.Span) {
				assert.Equal(1.0, span.Tag(ext.EventSampleRate))
			},
		},
                // ... continue
	} {
		assert := assert.New(t)
		conf := &api.Config{
			HttpClient: NewHTTPClient(tt.opts...),
			Address:    ts.URL,
		}
                 // ... perform some operation and get a span (like your getWriteSpan)
                 tt.test(assert, span)
        }
}
```
You won't need what is `getWriteSpans` anymore (it can be inlined), it will remove repetition completely, make tests small and setup and assertion isolated in the `[]struct` that we are ranging over. That way input and output is clear. What do you think?

Reply:
This is much better.